### DISCUSSION TITLE
Issue with applying kerning from this font

### DISCUSSION THREAD

**Discussion by schombert** [harfbuzz/harfbuzz#4751] [Q&A] [answered]
I am attempting to render text using the https://github.com/adobe-fonts/source-serif font via harfbuzz (with the smcp feature turned on), but it doesn't seem handle the kerning correctly. Prior to switching to harfbuzz, I was rendering text via freetype and manually getting kerning data out of the gpos table. This produced the following result, which I have confirmed with the font's creator is the intended result 
![266205700-ac6c9b57-d9fe-4f63-9674-19f0c681366b](https://github.com/harfbuzz/harfbuzz/assets/22087402/8429c822-4b29-47e5-9e77-e563f8002560)
This, however, is what I am currently getting via harfbuzz shaping and positioning glyphs according to the values in `hb_buffer_get_glyph_positions`
![image](https://github.com/harfbuzz/harfbuzz/assets/22087402/5480bd43-9ba8-48d2-a7d8-ca902cf89808)
As you can see, the kerning between the small caps A and small caps V doesn't seem to have been correctly applied. Is there a reference renderer or something along those lines that I can use to help verify whether this is an issue in my code or whether it really is a harfbuzz problem?

**Comment by behdad** [comment]
> Is there a reference renderer or something along those lines that I can use to help verify whether this is an issue in my code or whether it really is a harfbuzz problem?

Are you passing `HB_SCRIPT_LATIN` to HarfBuzz?

The reference renderers is `hb-view`. Also useful to inspect output of `hb-shape`.

**Reply by schombert** [reply]
I switched everything to a hard-coded
```
hb_buffer_set_direction(hb_buf, HB_DIRECTION_LTR);
hb_buffer_set_script(hb_buf, HB_SCRIPT_LATIN);
hb_buffer_set_language(hb_buf, hb_language_from_string("en", -1));
```
and the kerning issue persists

**Comment by schombert** [comment]
Also, is there a way to turn on `smcp` for hb-view and hb-shape? without that, I won't be able to check the issue.

**Reply by behdad** [reply]
> Also, is there a way to turn on `smcp` for hb-view and hb-shape? without that, I won't be able to check the issue.

`--features smcp`.

**Comment by schombert** [comment]
Ok, so after rendering it with hb-view, the kerning shows up correctly. So it is definitely something wrong I am doing; I just have no idea what 
![image](https://github.com/harfbuzz/harfbuzz/assets/22087402/99511f76-6cbb-4cbb-ac0a-efd9f6b2c6d6)

**Reply by behdad** [reply]
How are you opening the face? Is this with hb-ft or something?

**Reply by schombert** [reply]
via this
```
FT_New_Memory_Face(ft_library, fnt.file_data.get(), file_size, 0, &fnt.font_face);
FT_Select_Charmap(fnt.font_face, FT_ENCODING_UNICODE);
FT_Set_Pixel_Sizes(fnt.font_face, 0, dr_size);
fnt.hb_font_face = hb_ft_font_create(fnt.font_face, nullptr);
hb_font_set_scale(fnt.hb_font_face, dr_size, dr_size);
```

**Reply by behdad** [reply]
The problem is probably with dr_size. Try multiplying it by 64.

**Reply by behdad** [reply, marked as answer]
Or. Not call set_scale at all.

**Reply by schombert** [reply]
without set scale we get 
![image](https://github.com/harfbuzz/harfbuzz/assets/22087402/5d343838-814b-4dbc-9532-e19f4c518841)
which appears correct to me
Thank you so much for your time!

**Reply by behdad** [reply]
Looks good. Glad it worked out.
